### PR TITLE
Show derived value as QR code

### PR DESCRIPTION
**Pull request by angelix**
Match web [implementation](https://github.com/dicekeys/dicekeys-app-typescript/pull/223)

### CODE REVIEW

File: app/src/main/java/org/dicekeys/app/extensions/FragmentExtensions.kt [hunk 1]
```diff
@@ -103,4 +105,34 @@ fun Fragment.showPopupMenu(view: View, @MenuRes menuRes: Int, fn: ((popupMenu: P
     popup.setOnMenuItemClickListener(listener)
     fn?.invoke(popup)
     popup.show()
+}
+
+fun Fragment.askToCopyToClipboard(content: String, animateView: View? = null){
+    MaterialAlertDialogBuilder(requireContext())
+        .setTitle("Copy to Clipboard?")
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/extensions/FragmentExtensions.kt:112]: Is capitalizing Clipboard something that is recommended by the platform?

Comment (angelix) [reply app/src/main/java/org/dicekeys/app/extensions/FragmentExtensions.kt:112]: nope. changing it.

**Comment by UppaJung** [comment]
Very cool.  After spending a bunch of time staring at the code, I cannot figure out where you assemble the string that gets put in the QR code.  Is it just the string of characters or is it a URL with the string of characters as a parameter?  I see there's a parameter called "content" but where content is being set is...outside the PR?

**Comment by angelix** [comment]
> 



> Very cool. After spending a bunch of time staring at the code, I cannot figure out where you assemble the string that gets put in the QR code. Is it just the string of characters or is it a URL with the string of characters as a parameter? I see there's a parameter called "content" but where content is being set is...outside the PR?

Is the same content as the value in copy. Depending on the view, we get the value from `recipeViewModel.derivedValueAsString`.
```
recipeViewModel.derivedValueAsString.value?.let {
    dialogQR(title = recipeViewModel.derivedValueView.value!!.description, content = it)
}
```

**Comment by UppaJung** [comment]
> Is the same content as the value in copy. Depending on the view, we get the value from `recipeViewModel.derivedValueAsString`.

I see now.

Can we have another field which is derivedQrCodeTextAsString, which has a switch statement on the derivation type and defaults to just being an exact copy of derivedValueAString (the current behavior)?

For bips, do they not do something like bips39:/ ... as the format?

**Comment by angelix** [comment]
> Can we have another field which is derivedQrCodeTextAsString, which has a switch statement on the derivation type and defaults to just being an exact copy of derivedValueAString (the current behavior)?
> 
> For bips, do they not do something like bips39:/ ... as the format?

I can do that, but is going to be the exact copy right? We don't currently have a different output for any view.

For BIP39, a space separated list of words is the expected QR code.

**Comment by UppaJung** [comment]
> I can do that, but is going to be the exact copy right? We don't currently have a different output for any view.

No, it wouldn't always be an exact copy because if you don't have a scheme at the front of the text in a QR code to turn it into a URL, and you scan the QR code with the OS, it won't know which app to run.  If you put public keys in a URL, you can use a scheme to get the OS to launch an app that's appropriate for a key.
 
> For BIP39, a space separated list of words is the expected QR code.

Okay.  Are you sure the apps won't strip out a scheme in front?  It would make sense for them to do that so that the OS could launch the app if the user scanned it in via the OS (even if the app doesn't want the user doing this on purpose.)

**Comment by angelix** [comment]
> > I can do that, but is going to be the exact copy right? We don't currently have a different output for any view.
> 
> No, it wouldn't always be an exact copy because if you don't have a scheme at the front of the text in a QR code to turn it into a URL, and you scan the QR code with the OS, it won't know which app to run. If you put public keys in a URL, you can use a scheme to get the OS to launch an app that's appropriate for a key.
> 
> > For BIP39, a space separated list of words is the expected QR code.
> 
> Okay. Are you sure the apps won't strip out a scheme in front? It would make sense for them to do that so that the OS could launch the app if the user scanned it in via the OS (even if the app doesn't want the user doing this on purpose.)

Agree. We need to find or define schemes for the following:

```
class JSON : DerivedValueView("JSON")
class Password : DerivedValueView("Password")
class Hex : DerivedValueView("HEX")
class HexSigningKey : DerivedValueView("HEX (Signing Key)")
class HexUnsealing : DerivedValueView("HEX (Unsealing Key)")
class HexSealing : DerivedValueView("HEX (Sealing Key)")
class OpenPGPPrivateKey : DerivedValueView("OpenPGP Private Key")
class OpenSSHPrivateKey : DerivedValueView("OpenSSH Private Key")
class OpenSSHPublicKey : DerivedValueView("OpenSSH Public Key")
```

There is no spec about a scheme for BIP39 (or i couldn't find one). I know that the expected content is a plain string that you have to scan from the app itself. A scheme will break most of the apps if not all.

**Comment by angelix** [comment]
Don't really like the changes in 8a13d0308523829102dd621cbbf7cc45628ed9c6 to provide a URI based QR code as there is no spec for that. 

Also content for URI is url encoded.

adb_screenshot_1661518118

**Comment by UppaJung** [comment]
What is the scheme in the URI?  Not enough space to show the contents there?  I think it's worth showing the contents so that people know that when they point a camera at the URL what the camera can see is the contents.  (Otherwise, some people might think there's some special security in the QR code that encrypts it or something.)

**Comment by angelix** [comment]
> What is the scheme in the URI? Not enough space to show the contents there? I think it's worth showing the contents so that people know that when they point a camera at the URL what the camera can see is the contents. (Otherwise, some people might think there's some special security in the QR code that encrypts it or something.)

Please check screenshots [here](https://github.com/dicekeys/dicekeys-android/pull/202#issue-1336338369)

**Comment by UppaJung** [comment]
Looking great!  I was hoping to have a chance to perfect the wording in the TypeScript app.  I'll work on that.  Great to see it's quick to update this!

**Comment by UppaJung** [comment]
Improved warning text in https://github.com/dicekeys/dicekeys-app-typescript/pull/223/commits/6724b35dc0640a20f8504b685da3e91ff54f0ef3

I'd remove the URI option for now.  Among other things, it's not clear to the user what kind of URI it is, and it doesn't behave like the URIs they're familiar with (won't open a web app).

**Pull request merged**